src/rust/engine/src/intrinsics.rs

```diff
@@ -742,3 +752,85 @@ fn docker_resolve_image(
   }
   .boxed()
 }
+
+fn parse_python_deps(context: Context, args: Vec<Value>) -> BoxFuture<'static, NodeResult<Value>> {
+  in_workunit!(
+    "parse_python_dependencies",
+    Level::Debug,
+    |_workunit| async move {
+      let core = &context.core;
+      let directory_digest = Python::with_gil(|py| {
+        let py_digest = (*args[0]).as_ref(py);
+        lift_directory_digest(py_digest)
+      })?;
+
+      let mut path = None;
+      let mut digest = None;
+      directory_digest
+        .tree
+        .unwrap()
+        .walk(SymlinkBehavior::Oblivious, &mut |node_path, entry| {
+          if let Entry::File(file) = entry {
+            path = Some(node_path.to_owned());
+            digest = Some(file.digest());
+          }
+        });
+
+      let cache_key = CacheKey {
+        key_type: CacheKeyType::DepInferenceRequest.into(),
+        digest: Some(digest.unwrap().into()),
+      };
+      let cached_result = core.local_cache.load(&cache_key).await?;
+
+      let mut result: Option<ParsedPythonDependencies> = None;
+      if let Some(bytes) = cached_result {
+        result = serde_json::from_slice(&bytes).ok();
+      }
+      if result.is_none() {
+        let store = core.store();
+        let bytes = store
+          .load_file_bytes_with(digest.unwrap(), |bytes| Vec::from(bytes))
+          .await?;
+
+        let contents = std::str::from_utf8(&bytes)
+          .map_err(|err| format!("Failed to convert digest bytes to utf-8: {err}"))?;
+        result = Some(get_dependencies(contents, path.unwrap())?);
+        core
+          .local_cache
+          .store(
+            &cache_key,
+            Bytes::from(
+              serde_json::to_string(&result)
+                .map_err(|e| format!("Failed to serialize dep inference cache result: {e}"))?,
+            ),
+          )
+          .await?;
+      };
+      let result = result.unwrap();
```

Comment:
TODO: would replace the mutable result field usage with `match` or `if let` statements.

Reply:
I couldn't find the right touch for that. Let me try again